yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+
+  /** number of events to batch up before posting*/
+  private var batchSize: Int = DEFAULT_BATCH_SIZE
+
+  /** queue of actions*/
+  private val actionQueue = new LinkedBlockingQueue[QueuedAction]
+
+  /** cache layer to handle timeline client failure.*/
+  private var entityList = new LinkedList[TimelineEntity]
+
+  /** current entity; wil be created on demand. */
+  private var curEntity: Option[TimelineEntity] = None
+
+  /** Has a start event been processed? */
+  private val appStartEventProcessed = new AtomicBoolean(false)
+
+  /* has the application event event been processed */
+  private val appEndEventProcessed = new AtomicBoolean(false)
+
+  /** How many events have been received in the current entity */
+  private var currentEventCount = 0
+
+  /** counter of events processed -that is have been through handleEvent()*/
+  private val eventsProcessed: AtomicInteger = new AtomicInteger(0)
+
+  /** counter of events queued. */
+  private val eventsQueued: AtomicInteger = new AtomicInteger(0)
+
+  /** how many event postings failed? */
+  private val eventPostFailures: AtomicInteger = new AtomicInteger(0)
+
+  /** how many flushes have taken place? */
+  private val flushCount = new AtomicInteger(0)
+
+  /** Event handler */
+  private var eventHandlingThread: Option[Thread] = None
+
+  /**
+   * Flag to indicate the thread is stopped; events aren't being
+   * processed.
+   */
+  private val stopped: AtomicBoolean = new AtomicBoolean(true)
+
+  /**
+   * boolean to track whether a thread is active or not, for tests to
+   * monitor and see if the thread has completed.
+   */
+  private val postThreadActive: AtomicBoolean = new AtomicBoolean(false)
+
+  /**
+   * object used for a lock on entity operations
+   */
+  private val entityLock: AnyRef = new AnyRef
+
+  /**
+   * How long to wait for shutdown before giving up
+   */
+  private var maxTimeToWaitOnShutdown: Long = SHUTDOWN_WAIT_TIME
+
+  /** Domain ID for entities: may be null */
+  private var domainId: String = null
+
+  /** URI to timeline web application -valid after `serviceStart()` */
+  private var timelineWebappAddress: URI = _
+
+  /**
+   * Create a timeline client and start it. This does not update the
+   * `timelineClient` field, though it does verify that the field
+   * is unset.
+   *
+   * The method is private to the package so that tests can access it, which
+   * some of the mock tests do to override the timeline client creation.
+   * @return the timeline client
+   */
+  private [yarn] def createTimelineClient(): TimelineClient = {
+    require(timelineClient.isEmpty, "timeline client already set")
+    YarnTimelineUtils.createTimelineClient(sparkContext)
+  }
+
+  /**
+   * Get the timeline client.
+   * @return the client
+   * @throws Exception if the timeline client is not currently running
+   */
+  def getTimelineClient: TimelineClient = {
+    timelineClient.getOrElse(throw new Exception("Timeline client not running"))
```

Comment:
nit: I'm not a fan of throwing `Exception`. Either just call `timelineClient.get` or throw a better exception (e.g. `require(timelineClient.isDefined)`.